### PR TITLE
[canvaskit] Detect animated WebP images

### PR DESCRIPTION
**Pull request by harryterkelsen**
Reads the WebP header to determine if the WebP image is animated or not. If it's not animated, we can use `` tag decoding for less jank.

The WebP half of https://github.com/flutter/flutter/issues/151911

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/web_ui/lib/src/engine/image_format_detector.dart [hunk 1]
```diff
@@ -173,3 +183,80 @@ bool isAvif(Uint8List data) {
   }
   return false;
 }
+
+/// Reads the header of a WebP file to determine if it is animated or not.
+///
+/// See https://developers.google.com/speed/webp/docs/riff_container
+class WebpHeaderReader {
```
Comment (kevmoo) [inline lib/web_ui/lib/src/engine/image_format_detector.dart:190]: Nit: I kinda like when these types of things are private, in their own library and just the functional thing as a top-level function `bool isAnimated(bytes)`. With the class as private. One should only call `isAnimated` once, right?

File: lib/web_ui/test/engine/image_format_detector_test.dart [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        'blendBG.webp',
+        'required.webp',
+        'stoplight_h.webp',
+        'stoplight.webp',
+      ];
+
+      ImageType? expectedImageType;
+      final String testFileExtension =
+          testFile.substring(testFile.lastIndexOf('.') + 1);
+      switch (testFileExtension) {
```
Comment (kevmoo) [inline lib/web_ui/test/engine/image_format_detector_test.dart:64]: might could use switch expression here?

**Review by kevmoo — APPROVED** [review]

File: lib/web_ui/lib/src/engine/image_format_detector.dart (hunk 1 above)
Comment (harryterkelsen) [reply lib/web_ui/lib/src/engine/image_format_detector.dart:190]: Done. `isAnimated` isn't even exposed outside the file. It's an implementation detail of `detectImageType`

File: lib/web_ui/test/engine/image_format_detector_test.dart (hunk 2 above)
Comment (harryterkelsen) [reply lib/web_ui/test/engine/image_format_detector_test.dart:64]: Nice! Didn't know about this in the language

File: lib/web_ui/test/engine/image_format_detector_test.dart [hunk 3]
```diff
@@ -58,29 +58,20 @@ Future<void> testMain() async {
...
+        'jpg' => ImageType.jpeg,
+        'jpeg' => ImageType.jpeg,
+        'gif' => ImageType.animatedGif,
+        'webp' => animatedWebpFiles.contains(testFile)
+            ? ImageType.animatedWebp
+            : ImageType.webp,
+        'avif' => ImageType.avif,
+        'bmp' => ImageType.bmp,
+        'png' => ImageType.png,
+        String() => null,
```
Comment (kevmoo) [inline lib/web_ui/test/engine/image_format_detector_test.dart:73]: or

```suggestion
        _ => null,
```

🤷

File: lib/web_ui/lib/src/engine/image_format_detector.dart [hunk 4]
```diff
@@ -36,8 +36,8 @@ ImageType? detectImageType(Uint8List data) {
 
     // If we detect a WebP image, check if it is animated.
     if (format.imageType == ImageType.webp) {
-      final WebpHeaderReader webpHeaderReader =
-          WebpHeaderReader(data.buffer.asByteData());
+      final _WebpHeaderReader webpHeaderReader =
+          _WebpHeaderReader(data.buffer.asByteData());
       if (webpHeaderReader.isAnimated()) {
```
Comment (kevmoo) [inline lib/web_ui/lib/src/engine/image_format_detector.dart:41]:
```suggestion
      if (_WebpHeaderReader(data.buffer.asByteData()).isAnimated()) {
```

compromise?

Comment (kevmoo) [reply lib/web_ui/lib/src/engine/image_format_detector.dart:41]: meh...not a biggy. I'm being picky

Comment (harryterkelsen) [reply lib/web_ui/lib/src/engine/image_format_detector.dart:41]: done

File: lib/web_ui/test/engine/image_format_detector_test.dart (hunk 3 above)
Comment (harryterkelsen) [reply lib/web_ui/test/engine/image_format_detector_test.dart]: done

File: lib/web_ui/lib/src/engine/image_format_detector.dart [hunk 5]
```diff
@@ -173,3 +181,80 @@ bool isAvif(Uint8List data) {
...
+    // Read chunk size.
+    _readUint32();
+    return chunkFourCC == expectedHeader;
+  }
+
+  /// Reads the WebP header. Returns [false] if this is not a valid WebP header.
+  bool _readWebpHeader() {
+    final String riffBytes = _readFourCC();
+
+    // Read file size byte.
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/image_format_detector.dart:229]: Only one byte? 🙂

**Review by mdebbar — COMMENTED** [review]
Sorry for the late review, but this LGTM too!

File: lib/web_ui/lib/src/engine/image_format_detector.dart (hunk 5 above)
Comment (kevmoo) [reply lib/web_ui/lib/src/engine/image_format_detector.dart:229]: get out of here with your snark! 😜

**Pull request merged**